src/ot/layout/common.rs

```diff
@@ -0,0 +1,835 @@
+};
+use ttf_parser::GlyphId;
+
+use crate::font::Font;
+use crate::Tag;
+
+/// A GSUB or GPOS table.
+#[derive(Clone, Copy, Debug)]
+pub struct SubstPosTable<'a> {
+    scripts: RecordList<'a>,
+    features: RecordList<'a>,
+    lookups: LookupList<'a>,
+    variations: Option<FeatureVariations<'a>>,
+}
+
+#[allow(dead_code)]
+impl<'a> SubstPosTable<'a> {
+    pub fn parse(data: &'a [u8]) -> Option<Self> {
+        let mut s = Stream::new(data);
+
+        let major_version = s.read::<u16>()?;
+        let minor_version = s.read::<u16>()?;
+        if major_version != 1 {
+            return None;
+        }
+
+        let scripts = RecordList::parse(s.read_at_offset16()?)?;
+        let features = RecordList::parse(s.read_at_offset16()?)?;
+        let lookups = LookupList::parse(s.read_at_offset16()?)?;
+
+        let mut variations = None;
+        if minor_version >= 1 {
+            variations = FeatureVariations::parse(s.read_at_offset32()?);
+        }
+
+        Some(Self { scripts, features, lookups, variations })
+    }
+
+    // Scripts
+
+    pub fn script_count(&self) -> u16 {
+        self.scripts.len()
+    }
+
+    pub fn get_script(&self, index: ScriptIndex) -> Option<Script<'a>> {
+        Script::parse(self.scripts.get_data(index.0)?)
+    }
+
+    pub fn get_script_tag(&self, index: ScriptIndex) -> Option<Tag> {
+        self.scripts.get_tag(index.0)
+    }
+
+    pub fn find_script_index(&self, script_tag: Tag) -> Option<ScriptIndex> {
+        self.scripts.find_index(script_tag).map(ScriptIndex)
+    }
+
+    // Features
+
+    pub fn feature_count(&self) -> u16 {
+        self.features.len()
+    }
+
+    pub fn get_feature(&self, index: FeatureIndex) -> Option<Feature<'a>> {
+        Feature::parse(self.features.get_data(index.0)?)
+    }
+
+    pub fn get_feature_tag(&self, index: FeatureIndex) -> Option<Tag> {
+        self.features.get_tag(index.0)
+    }
+
+    pub fn find_feature_index(&self, feature_tag: Tag) -> Option<FeatureIndex> {
+        self.features.find_index(feature_tag).map(FeatureIndex)
+    }
+
+    // Lookups
+
+    pub fn lookup_count(&self) -> u16 {
+        self.lookups.len()
+    }
+
+    pub fn get_lookup(&self, index: LookupIndex) -> Option<Lookup<'a>> {
+        self.lookups.get(index)
+    }
+
+    // Feature variations
+
+    pub fn variation_count(&self) -> u32 {
+        self.variations.map_or(0, |var| var.len())
+    }
+
+    pub fn get_variation(
+        &self,
+        feature_index: FeatureIndex,
+        var_index: VariationIndex,
+    ) -> Option<Feature<'a>> {
+        self.variations
+            .and_then(|var| var.find_substitute(feature_index, var_index))
+            .or_else(|| self.get_feature(feature_index))
+    }
+
+    pub fn find_variation_index(&self, coords: &[i32]) -> Option<VariationIndex> {
+        self.variations?.find_index(coords)
+    }
+}
+
+/// A type-safe wrapper for a script index.
```

Comment:
Isn't this is the same code as in `ttf_parser::ggg`?

Reply:
Um, yes, mostly. My version is more analogous to harfbuzz's (not iterator style). For a public API yours is probably nicer, but for integration with harfbuzz this way it was a bit simpler because it provides exactly the functions harfbuzz had and needs. Also a problem would have been that ttf_parser uses `NormalizedCoordinate` for feature variations, but harfbuzz uses `i32`. And the non-iterator style can use binary search sometimes, but that's a minor thing. Ultimately, both ways would have worked out, I guess.

One thing I noticed while looking at `ttf_parser`'s source code: I think you're `Iterator::nth` implementations are incorrect. In your implementation `nth` behaves like an array access, but afaik `nth` is consuming, i.e. `nth(0)` is equivalent to `next()` and calling it multiple times should produce different elements.

Reply:
Ok. Then we can remove it from `ttf_parser`. It was just an experimental implementation of mine. 

I will look into `nth()` implementation.